src/Streamer.cpp

```diff
@@ -43,102 +42,133 @@ FileWriter::Streamer::connect(std::string TopicName) {
+      LOG(Sev::Info, "Stop stream timeout for topic \"{}\" reached. {} ms "
+                     "passed since last message received.",
+          MessageProcessor.topic(),
+          (systemTime() - LastMessageTimestamp).count());
       Sources.clear();
-      return ProcessMessageResult::STOP();
+      return ProcessMessageResult::STOP;
     }
-    return ProcessMessageResult::OK();
+    return ProcessMessageResult::OK;
   }
   if (Poll.isErr()) {
-    return ProcessMessageResult::ERR();
+    return ProcessMessageResult::ERR;
   }
 
   // convert from KafkaW to Msg
-  Msg Message(Msg::fromKafkaW(Poll.isMsg()));
-  if (Message.type == MsgType::Invalid) {
-    return ProcessMessageResult::ERR();
+  auto KafkaMessage = Poll.isMsg();
+  std::unique_ptr<FlatbufferMessage> Message;
+  try {
+    Message.reset(new FlatbufferMessage(
+        reinterpret_cast<const char *>(KafkaMessage->data()),
+        KafkaMessage->size()));
+  } catch (std::runtime_error &Error) {
+    LOG(Sev::Warning,
+        "Message that is not a valid flatbuffer encountered. The error was: {}",
+        Error.what());
+    return ProcessMessageResult::ERR;
   }
 
-  // Make sure that the message source is relevant and that the message is in
-  // the correct time window
-  DemuxTopic::DT MessageTime =
-      MessageProcessor.time_difference_from_message(Message);
-  if (std::find(Sources.begin(), Sources.end(), MessageTime.sourcename) ==
+  LastMessageTimestamp = systemTime(); // For dealing with messsage timeouts
+
+  if (std::find(Sources.begin(), Sources.end(), Message->getSourceName()) ==
       Sources.end()) {
-    return ProcessMessageResult::OK();
+    LOG(Sev::Warning, "Message from topic\"{}\" has an unknown source name "
+                      "(\"{}\"), ignoring.",
+        MessageProcessor.topic(), Message->getSourceName());
+    return ProcessMessageResult::OK;
   }
-  if (MessageTime.dt < std::chrono::duration_cast<std::chrono::nanoseconds>(
-                           Options.StartTimestamp)
-                           .count()) {
-    return ProcessMessageResult::OK();
+
+  // Timestamp of message is before the "start" timestamp
+  if (static_cast<std::int64_t>(Message->getTimestamp()) <
+      std::chrono::duration_cast<std::chrono::nanoseconds>(
+          Options.StartTimestamp)
+          .count()) {
+    return ProcessMessageResult::OK;
   }
+
+  // Check if there is a stop stime configured and the message timestamp is
+  // greater than it
   if (Options.StopTimestamp.count() > 0 &&
-      MessageTime.dt > std::chrono::duration_cast<std::chrono::nanoseconds>(
-                           Options.StopTimestamp)
-                           .count()) {
-    if (removeSource(MessageTime.sourcename)) {
-      return ProcessMessageResult::STOP();
+      static_cast<std::int64_t>(Message->getTimestamp()) >
+          std::chrono::duration_cast<std::chrono::nanoseconds>(
+              Options.StopTimestamp)
+              .count()) {
```

Comment:
bool stoptimeElapsed(std::uint64_t Timestamp, std::chrono::milliseconds Stoptime)

Reply:
The timing code is an excellent candidate for removing from the `Streamer` class and put somewhere else in order to simplify unit testing and modularise the code a bit. Until then, I have done as you suggested.